emu_sv/custom_callback_implementations.py

```diff
@@ -1,93 +1,117 @@
-        qubit_density[i] = state_tensor.select(i, 1).norm() ** 2
-    return qubit_density
+        state_tensor = state.vector.view(2**i, 2, -1)
+        # nᵢ is a projector and therefore nᵢ == nᵢnᵢ
+        # ❬ψ|nᵢ|ψ❭ == ❬ψ|nᵢnᵢ|ψ❭ == ❬ψ|nᵢ * nᵢ|ψ❭ == ❬ϕ|ϕ❭ == |ϕ|**2
+        occupation[i] = torch.linalg.vector_norm(state_tensor[:, 1]) ** 2
+    return occupation
 
 
 def correlation_matrix_sv_impl(
     self: CorrelationMatrix,
-    config: BackendConfig,
-    t: int,
+    *,
+    config: EmulationConfig,
     state: StateVector,
-    H: Operator,
+    hamiltonian: DenseOperator,
 ) -> torch.Tensor:
     """
-    Custom implementation of the density-density correlation ❬ψ|nᵢnⱼ|ψ❭ for the state vector solver.
-
+    Custom implementation of the density-density correlation ❬ψ|nᵢnⱼ|ψ❭
+      for the state vector solver.
     TODO: extend to arbitrary two-point correlation ❬ψ|AᵢBⱼ|ψ❭
     """
-    nqubits = int(math.log2(len(state.vector)))
-    state_tensor = state.vector.reshape((2,) * nqubits)
-
+    nqubits = state.n_qudits
     correlation = torch.zeros(
-        nqubits, nqubits, dtype=torch.float64, device=state_tensor.device
+        nqubits, nqubits, dtype=torch.float64, device=state.vector.device
     )
 
     for i in range(nqubits):
-        select_i = state_tensor.select(i, 1)
+        select_i = state.vector.view(2**i, 2, -1)
+        select_i = select_i[:, 1]
         for j in range(i, nqubits):  # select the upper triangle
             if i == j:
-                value = select_i.norm() ** 2
+                value = torch.linalg.vector_norm(select_i) ** 2
+                correlation[j, j] = value
             else:
-                value = select_i.select(j - 1, 1).norm() ** 2
+                select_i = select_i.view(2**i, 2 ** (j - i - 1), 2, -1)
+                select_ij = select_i[:, :, 1, :]
+                value = torch.linalg.vector_norm(select_ij) ** 2
+                correlation[i, j] = value
+                correlation[j, i] = value
 
-            correlation[i, j] = value
-            correlation[j, i] = value
     return correlation
 
 
 def energy_variance_sv_impl(
     self: EnergyVariance,
-    config: BackendConfig,
-    t: int,
+    *,
+    config: EmulationConfig,
     state: StateVector,
-    H: RydbergHamiltonian,
+    hamiltonian: RydbergHamiltonian,
 ) -> torch.Tensor:
     """
     Custom implementation of the energy variance ❬ψ|H²|ψ❭-❬ψ|H|ψ❭² for the state vector solver.
     """
-    hstate = H * state.vector
-    h_squared = torch.vdot(hstate, hstate).real
-    energy = torch.vdot(state.vector, hstate).real
-    energy_variance: torch.Tensor = h_squared - energy**2
-    return energy_variance
+    hstate = hamiltonian * state.vector
+    h_squared = torch.vdot(hstate, hstate)
+    energy = torch.vdot(state.vector, hstate)
+    en_var: torch.Tensor = h_squared - energy**2
+    assert torch.allclose(en_var.imag, torch.zeros_like(en_var.imag), atol=1e-8)
+    return en_var.real
```

Comment:
it's better to cast to real early, it saves flops in squaring and adding, although given that this is python, maybe it's outweighed by the fact that you reference one fewer attribute. 

Also, why the assert?

Reply:
I had in mind some error accumulation in the state with time evolution. Such that positive observables acquire imaginary part.

Now I see it's an overkill. I'm going to remove this unnecessary checks